### PR TITLE
Implement the new installation and configuration mechanism

### PR DESCRIPTION
**Pull request by GLinnik21**
As suggested by @naftaly in https://github.com/kstenerud/KSCrash/issues/491#issuecomment-2132267462

### CODE REVIEW

File: Sources/KSCrashRecording/include/KSCrashConfiguration.h [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+/** Callback to invoke upon a crash.
+ *
+ * This function is called during the crash reporting process, providing an opportunity
+ * to add additional information to the crash report. Only async-safe functions should
+ * be called from this function. Avoid calling Objective-C methods.
+ *
+ * **Default**: NULL
+ */
+@property (nonatomic, copy, nullable) void (^crashNotifyCallback)(const struct KSCrashReportWriter* writer);
```
Comment (GLinnik21) [inline Sources/KSCrashRecording/include/KSCrashConfiguration.h:87]: This is the "ugly" API for Swift. `KSCrashReportWriter` is a struct that deeply integrated in KSCrash and used in callback, that makes it hard to proxy. 
Visible to Swift:
```swift
open var crashNotifyCallback: ((UnsafePointer<KSCrashReportWriter>) -> Void)?
```

Comment (bamx23) [reply Sources/KSCrashRecording/include/KSCrashConfiguration.h:87]: Wrapping this in an ObjC class would be a really good thing to have. Swift API matters.

File: Sources/KSCrashRecording/include/KSCrashCConfiguration.h [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+     * C++ exceptions by swapping the `__cxa_throw` function. It helps in obtaining
+     * accurate stack traces even in dynamically linked libraries and allows overriding
+     * the original `__cxa_throw` with a custom implementation.
+     *
+     * **Default**: false
+     */
+    bool enableSwapCxaThrow;
+} KSCrashCConfiguration;
+
+#define KSCrashConfig_Default (KSCrashCConfiguration) { \
```
Comment (bamx23) [inline Sources/KSCrashRecording/include/KSCrashCConfiguration.h:143]: Why is it a define (and not let's say a function)?

File: Sources/KSCrashRecording/include/KSCrashConfiguration.h (hunk 1 above)
Comment (GLinnik21) [reply Sources/KSCrashRecording/include/KSCrashConfiguration.h:87]: From docs
```
 * WARNING: Only call async-safe functions from this function! DO NOT call
 * Objective-C methods!!!
 ```

Comment (bamx23) [reply Sources/KSCrashRecording/include/KSCrashConfiguration.h:106]: Right. Well, it's fine then.
I'm thinking that a wrapper with all methods being `__attribute__((objc_direct))` might do the trick, but it's still ObjC.

File: Sources/KSCrashRecording/include/KSCrash.h [hunk 3]
```diff
@@ -231,6 +103,9 @@ typedef enum
 
 #pragma mark - API -
 
+/** Init KSCrash instance with custom base path. */
+- (id) initWithBasePath:(NSString*) basePath;
```
Comment (bamx23) [inline Sources/KSCrashRecording/include/KSCrash.h:107]: I see that you're copying the method from other place, but why it's not `instancetype` here?

File: Sources/KSCrashRecording/KSCrash.m [hunk 4]
```diff
@@ -325,14 +249,10 @@ - (NSDictionary*) systemInfo
...
+- (BOOL) installWithConfiguration:(KSCrashConfiguration*) configuration
 {
-    _monitoring = kscrash_install(self.bundleName.UTF8String,
-                                          self.basePath.UTF8String);
-    if(self.monitoring == 0)
-    {
-        return false;
-    }
+    self.configuration = [configuration copy];
+    kscrash_install(self.bundleName.UTF8String, self.basePath.UTF8String, [self.configuration toCConfiguration]);
```
Comment (bamx23) [inline Sources/KSCrashRecording/KSCrash.m:255]: What if `self.configuration == nil`? Maybe it's better to have a function that accepts a potentially nil ObjC config and returns structure (default if nil)

File: Sources/KSCrashRecording/KSCrashConfiguration.m [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        _userInfoJSON = nil;
+        _deadlockWatchdogInterval = 0.0;
+        _enableQueueNameSearch = NO;
+        _enableMemoryIntrospection = NO;
+        _doNotIntrospectClasses = nil;
+        _crashNotifyCallback = nil;
+        _reportWrittenCallback = nil;
+        _addConsoleLogToReport = NO;
+        _printPreviousLogOnStartup = NO;
+        _maxReportCount = 5;
```
Comment (bamx23) [inline Sources/KSCrashRecording/KSCrashConfiguration.m:48]: As this is a wrapper for C struct, maybe it's better to use defaults defined there (e.g. by creating a default struct here)? As it's an easy way to get inconsistency later. And sure I'm not asking reverse mapping of arrays (that are empty) and wrapping callbacks. Having several constants for non-zero values would also work (at least suggest editing in multiple places later)

**Review by bamx23 — CHANGES_REQUESTED** [review]

File: Sources/KSCrashRecording/KSCrash.m (hunk 4 above)
Comment (GLinnik21) [reply Sources/KSCrashRecording/KSCrash.m:255]: I think `self.configuration = [configuration copy] ?: [KSCrashConfiguration new];` would be enough, as `kscrash_install` accepts struct, not a pointer.

File: Sources/KSCrashRecording/KSCrashConfiguration.m [hunk 6]
```diff
@@ -35,18 +35,35 @@ - (instancetype)init
...
+                _userInfoJSON = nil;  // Handle the error appropriately
+            }
+        }
+        else
+        {
+            _userInfoJSON = nil;
+        }
+
+        _deadlockWatchdogInterval = cConfig.deadlockWatchdogInterval;
+        _enableQueueNameSearch = cConfig.enableQueueNameSearch ? YES : NO;
```
Comment (bamx23) [inline Sources/KSCrashRecording/KSCrashConfiguration.m:57]: Is there a reason for ternary operator here and below?
```suggestion
        _enableQueueNameSearch = (BOOL)cConfig.enableQueueNameSearch;
```

File: Sources/KSCrashRecording/KSCrash.m (hunk 4 above)
Comment (bamx23) [reply Sources/KSCrashRecording/KSCrash.m:255]: Yep, that's good too

**Review by bamx23 — APPROVED** [review]
Thanks for addressing the comments. This new way of initialization makes perfect sense here. And as mentioned in another PR, let's update README later to have actual usage instructions.

File: Sources/KSCrashRecording/KSCrashConfiguration.m (hunk 6 above)
Comment (GLinnik21) [reply Sources/KSCrashRecording/KSCrashConfiguration.m:57]: Because `YES` is `1`, but `true` is `!1`. I think cast wouldn't fix it.

**Pull request merged**